### DISCUSSION TITLE
gdl output through nohup delayed

### DISCUSSION THREAD

**Discussion by brandy125** [gnudatalanguage/gdl#1346] [Q&A] [answered]
I am used to spawn idl programs with nohup as:

spawn,"nohup idl <<< printloop &"

where printloop.pro could be 
```
pro printloop

while 1 do begin
	print,systime()
	wait,1
endwhile

end
```

To keep it simple we do not need the spawn here. So I compare

>nohup idl <<< printloop 

with 

>nohup gdl <<< printloop 

In the first case I get the time printed every second to the file nohup.out whereas in the second case it takes many seconds until a bunch of lines are being printed to nohup.out.

Where does this big difference come from? Why the output of gdl is soo slow compared to idl?

Anybody knows?

**Comment by slayoo** [comment]
Thanks for the report. I've just tried with freshly compiled `gdl` from the Github master branch and (trying out the simpler case without spawn):
- I get one output line every second (visible in the terminal)
- nothing is written to `nohup.out` (no such file is created)
 
Reading through `nohup` man page, I've noticed:
> NOTE: your shell may have its own version of nohup, which usually supersedes the  version  described  here.   Please  refer  to your shell's documentation for details about the options it supports.

Which version of GDL, shell and nohup are you using?

**Comment by brandy125** [comment]
Thank you for verifying!
I tested on 

```
GDL> !version
{
    "ARCH": "aarch64",
    "OS": "linux",
    "OS_FAMILY": "unix",
    "OS_NAME": "linux",
    "RELEASE": "8.2",
    "BUILD_DATE": "Jul 27 2022",
    "MEMORY_BITS": 64,
    "FILE_OFFSET_BITS": 64
}
GDL> !gdl
{
    "RELEASE": "1.0.1 Git",
    "BUILD_DATE": "Jul 27 2022",
    "EPOCH": 1658890800,
    "GDL_USE_DSFMT": 1,
    "GDL_USE_WX": 1,
    "GDL_POSIX": 1
}
GDL> 
```

and 

```
GDL>  !version
{
    "ARCH": "x86_64",
    "OS": "linux",
    "OS_FAMILY": "unix",
    "OS_NAME": "linux",
    "RELEASE": "8.2",
    "BUILD_DATE": "Apr 21 2022",
    "MEMORY_BITS": 64,
    "FILE_OFFSET_BITS": 64
}
GDL>  !gdl
{
    "RELEASE": "1.0.1 Git",
    "BUILD_DATE": "Apr 21 2022",
    "EPOCH": 1650510000,
    "GDL_USE_DSFMT": 1,
    "GDL_USE_WX": 1,
    "GDL_POSIX": 1
}
```

The program /usr/bin/nohup comes from 

```
coreutils-9.1-2.3.aarch64  or coreutils-8.32-8.5.x86_64
```

with 

```
bash-5.1.8-2.1.x86_64 or bash-5.1.16-8.3.aarch64
```

Can I try anything else?

**Comment by GillesDuvert** [comment]
Is that a problem with GDL's readline facilities? Anyway, worksforme:
Below Mageia 8 (5.15.55-desktop-2.mga8) , looks good..
```
GDL> !version
{
    "ARCH": "x86_64",
    "OS": "linux",
    "OS_FAMILY": "unix",
    "OS_NAME": "linux",
    "RELEASE": "8.2",
    "BUILD_DATE": "Jul 13 2022",
    "MEMORY_BITS": 64,
    "FILE_OFFSET_BITS": 64
}
```
```
$ nohup gdl -e "while 1 do begin & print, systime() & wait,1 & endwhile"
nohup: ignoring input and appending output to 'nohup.out'
^C$ 
$ cat nohup.out 
% PREF_SET: Unknown preference: IDL_RBUF_SIZE
% PREF_SET: Unknown preference: IDL_MORE
Mon Aug  1 15:50:01 2022
Mon Aug  1 15:50:02 2022
Mon Aug  1 15:50:03 2022
Mon Aug  1 15:50:04 2022
Mon Aug  1 15:50:05 2022
Mon Aug  1 15:50:06 2022
Mon Aug  1 15:50:07 2022
Mon Aug  1 15:50:08 2022
Mon Aug  1 15:50:09 2022
Mon Aug  1 15:50:10 2022
Mon Aug  1 15:50:11 2022
Mon Aug  1 15:50:12 2022
Mon Aug  1 15:50:13 2022
Mon Aug  1 15:50:14 2022
Mon Aug  1 15:50:15 2022
Mon Aug  1 15:50:16 2022
Mon Aug  1 15:50:17 2022
% Interrupt encountered.
% Interrupted at: $MAIN$   
```

**Comment by brandy125** [comment]
Try to check the output of nohup.out in a second shell while the program is still running. 
In my case the file is not updating until I press CTRL+C in the program. At this moment all lines appear in the nohup.out.

**Reply by GillesDuvert** [reply]
OK, sorry for the misunderstanding.
I would says that GDL 'print' does not flush the c++ ostream after write.

**Comment by brandy125** [comment]
Should/Could we change this behaviour?

**Comment by GillesDuvert** [comment]
solved in  #1348

**Comment by brandy125** [comment, marked as answer]
Very nice! Works exactly as in IDL now!

Thanks!
